### ISSUE TITLE
Serious Question - Porting to new distro - Procedure

### ISSUE THREAD

**Issue by monsieuremre** [Kicksecure/security-misc#162]
Since I have previously made the suggestion to base kicksecure to Tumbleweed, I can't unsee all the benefits everywhere. Our suid disabler is one of them. The Qubes vulnurability ticket. The package that causes the vulnurability to escelate privileges to root, is whitelisted in our hand written service.

On Tumbleweed, we would just do suid whitelisting on the default shipped secure file permissions configuration. See [here](https://github.com/openSUSE/permissions/blob/master/profiles/permissions.secure).

I also can't unsee the fact that we maintain our service that is not as efficient and we also have to play along with dpkg seperately so that our secure permissions don't get overridden. These are all automatically managed in Tumbleweed.

In most cases, a Tumbleweed base would reduce the complexity of Kicksecure from maintaining a lot of services to maintaining a lot of mere configuration files.

What is the procedure to switch to Tumbleweed? It seems kicksecure is tied too much to debian. What packages need to be ported, very specifically. Which ones. Tell me. Even if you don't want to port at all, you do you, I still want to try for myself to port. I want to try to create a Kicksecure SUSE edition, or whatever I'm allowed call it while respecting all the trademarks.

Someone has to do this to prove it possible. I did research. Even tho challenging in the beginning, I sincerely believe Tumbleweed will make development and maintaining easier in the long run, not to mention all the security benefits already in place in openSUSE Tumbleweed.

**Comment by monsieuremre** [comment]
It can also reduce maintaining an ISO image to maintaining a file. I'm now trying to create an opensuse iso that would install very secure defaults. I wouldn't be able to do this easily with other distros. See [here](https://doc.opensuse.org/projects/autoyast/). Works for Tumbleweed too.

**Comment by roddhjav** [comment]
They have been multiple (more or less serious) posts regarding porting Whonix to X on the whonix forum... However, I have to say that it is the first time that I see actual value in the proposition, as yes in the current setup, Tumbleweed is a better base distribution that Debian (I have no power in the decision anyway).

Regarding the porting tasks itself

- How well do you know dpkg & rpm packages. As all packages would need to be converted, there is a **lot** of work. Also, `config-package-dev` is used quite a lot, I do not know the rpm equivalent.
- The VM images are built with https://github.com/Kicksecure/derivative-maker, the `packages` include all specials pkg built with the distribution. You may have to add support for rpm in derivative-maker or create a new build system closer to the opensuse way of doing image build.

Last, but not the least. Nowadays all new [hardened/secure](https://clip-os.org/en/) OS aim to implement an immutable structure, so instead of fixing a security architecture from the 90', it could be interesting to move a more modern and state of the art immutable architecture... In other words, while keeping opensuse, did you consider basing the port on [MicroOS](https://microos.opensuse.org/)?

**Comment by adrelanos** [comment]
The first and rough steps I documented earlier here:
https://www.whonix.org/wiki/Dev/Operating_System#Requesting_a_Port_to_a_Different_Base_Operating_System

If these first steps were done, that would be useful.

> The Qubes vulnurability ticket. The package that causes the vulnurability to escelate privileges to root, is whitelisted in our hand written service.

Answered here:
https://github.com/Kicksecure/security-misc/issues/163

> On Tumbleweed, we would just do suid whitelisting on the default shipped secure file permissions configuration. See [here](https://github.com/openSUSE/permissions/blob/master/profiles/permissions.secure).
> 
> I also can't unsee the fact that we maintain our service that is not as efficient and we also have to play along with dpkg seperately so that our secure permissions don't get overridden. These are all automatically managed in Tumbleweed.
> 
> In most cases, a Tumbleweed base would reduce the complexity of Kicksecure from maintaining a lot of services to maintaining a lot of mere configuration files.

Maybe but such a port is no piece of cake.

> What is the procedure to switch to Tumbleweed?

> It seems kicksecure is tied too much to debian.

Hard to avoid.

> What packages need to be ported, very specifically.

There are packages here;

* https://github.com/orgs/Kicksecure/repositories?type=all
* https://github.com/orgs/Whonix/repositories?type=all (optional if this question only about Kicksecure, not Whonix)
* except for:
  * wiki backups (which isn't really software)
  * grml-debootstrap (could be replaced with a respective tool of the new distribution)

I don't consider any of these superfluous. Otherwise these wouldn't exist.

Porting [genmkfile](https://github.com/Kicksecure/genmkfile) to be able to handle RPM, maybe (mostly) automate RPM specfile generation would go a long way.

> Which ones. Tell me.

I could also say, finding out which packages to port and which ones aren't needed is a task assigned to the porter.

> I want to try to create a Kicksecure SUSE edition, or whatever I'm allowed call it while respecting all the trademarks.

> Even if you don't want to port at all, you do you, I still want to try for myself to port.

> Someone has to do this to prove it possible.

Right. Don't let me or and discussions slow you down. I couldn't have created Whonix by trying to find consensus with Tails and attempting to modify Tails so it (optionally) becomes similar to Whonix. Such fundamental design changes rarely happen as far as I know. Reference:
https://www.whonix.org/wiki/Dev/Operating_System#Tails

Sometimes you need to jump right into coding action to get an idea of the scope and feasibility to reach a proof of concept.

It however would be nice if most of the work was mergeable upstream at derivative-maker, Kicksecure (and Whonix) to have at least a (mostly) shared codebase.

If some difficult chunks were done then this would look much more realistic.

related:

* https://github.com/Kicksecure/security-misc/issues/138

In summary, some specific first steps:

* add new distribution to the comparison table
* create the RPM specfiles for at least 1 package
* add RPM support to genmkfile
* (mostly) automate RPM specfile creation (in genmkfile or developer-meta-files)

Next steps:

* possibility to configure trademarks in derivative-maker / packages (default: no trademark, opt-in for builders)
* add RPM specfile to all required packages
* port derivative-maker to new distribution

**Comment by monsieuremre** [comment]
Porting derivative maker is the tricky part. This package is more or less universal for example. Cut out the apt stuff and write a rpm-spec file and you are good to go. Derivative maker however... Hmmm. As Pujol suggested MicroOS would be an even better choice. It is pretty much Tumbleweed, but immutable. But making a modified ISO just as easy. See [here](https://opensuse.github.io/fuel-ignition/). It is a GUI tool to pre-define and create your own immutable MicroOS ISO basically. My point is not just that it is better, it is easier to maintain.

I've now learned how to package RPM's and I can write specs and makefiles.
> add RPM support to genmkfile
But this, I don't know if I can do. But my understanding is, rpms are really not that different. It is pretty much just the manifestation and the spec file, which are also similar enough. So Imma see what can be done.

**Comment by monsieuremre** [comment]
I found a way. A rather [simple one](https://github.com/jordansissel/fpm). I am going to try it. If it works, we would have to just maintain a tarball with the directory tree and stuff. Easy peasy. https://fpm.readthedocs.io/en/latest/getting-started.html#configuration-file

Can you please examine before I write a mechanism to make both rpm and deb from a directory alone. Is there any features missing here that you think is necessary? There are no post and pre install scripts for example but apparently it is not needed, because it creates them automatically, reading the makefile.

**Comment by adrelanos** [comment]
The `genmkfile install` target might be useful.

----

derivative-maker has quite a lot features:
https://www.kicksecure.com/wiki/Dev/Derivative-Maker

----

> Porting derivative maker is the tricky part.

And upon reflection, not that important. Not even the hardest part of me. Maybe. The build script could be (initially) in a separate repository or in a standalone/different folder inside derivative-maker. Once I see the scripts, configs I might be able to integrate it.

> It is a GUI tool to pre-define and create your own immutable MicroOS ISO basically.

A GUI tool is OK to initially create a config but for others who want to re-build from source code it should be possible to build using scripts/configs only without having to follow instructions "press this and that button in the GUI".

> My point is not just that it is better, it is easier to maintain.

That remains to be seen. Once you get started, you'll bump into various complexities trying to port. Already mentioned: config-package-dev. Also I think /usr/share/pam-configs might be a Debian derivatives only thing. For starters, porting that is not easy.

> I found a way. A rather [simple one](https://github.com/jordansissel/fpm). I am going to try it. If it works, we would have to just maintain a tarball with the directory tree and stuff. Easy peasy. https://fpm.readthedocs.io/en/latest/getting-started.html#configuration-file

First potential issue: https://fpm.readthedocs.io/en/latest/installation.html#installing-fpm shows `gem install fpm` which then doesn't verify gpg signatures? That is usually the issue with the fancy package managers (like NPM, large dependency trees, sometimes malicious dependencies, and no gpg verification).

> Can you please examine before I write a mechanism to make both rpm and deb from a directory alone. Is there any features missing here that you think is necessary? There are no post and pre install scripts for example but apparently it is not needed, because it creates them automatically, reading the makefile.

There's too much functionality in the various postinst scripts. Have a look. Could use something like this (adjust editor).

    find . -type f -not -iwholename '*.git*' | grep postinst | xargs editor

Well... Try... Then use `lintian --pedantic` (a deb package linter) and compare the diff of dpkg-buildpackage (or cowbuilder) created package vs fpm created package using `diffoscope`.

**Comment by monsieuremre** [comment]
Well I did manage to port it, but there are many changes. I'll go ahead and say packaging rpm's is easier. I got it working like a charm, everything is just like one file. pre and post scripts are in there too. It took me a lot longer to make a deb building mechanism. I couldn't manage to make a system that can build both rpm and deb. Because the rpm version uses a significantly different base. Apt stuff and some services are completely gone, because they are no longer needed in suse. This was just writing a makefile. I didn't do any of the genmakefile stuff. I wrote it manually for this package.

**Comment by adrelanos** [comment]
* https://forums.whonix.org/t/nixos-distro-preview/19883/2
* https://www.whonix.org/wiki/Dev/Operating_System#NixOS
